src/toplevel.pl

```diff
@@ -200,9 +203,144 @@
+    % if we are still inside the query or not.
+    '$get_b_value'(B0),
+    catch(call_residue_vars(user:QueryTerm, ResVars), Exception, Excepted = true),
+    gather_query_vars(VarNames, Vars0),
+    '$term_variables_under_max_depth'(Vars0, 22, Vars1),
+    '$project_atts':project_attributes(Vars1, ResVars),
+    '$get_b_value'(B),
+    (   B0 == B ->
+        % We are out of the choicepoint, ignore tail false
+        !
+    ;   Pending = true
+    ),
+    (   Excepted == true ->
+        !,
+        call(Callback_1, final(exception(Exception)))
+    ;   (   VarNames == [], ResGoals == [] ->
+            (   Pending == true ->
+                call(Callback_1, pending(true))
+            ;   call(Callback_1, final(true))
+            )
+        ;   copy_term([Vars1, ResVars], [Vars1, ResVars], ResGoals),
+            term_variables(ResGoals, ResGoalVars),
+            append([Vars1, ResGoalVars, ResVars], Vars2),
+            charsio:extend_var_list(Vars2, VarNames, NewVarNames, fabricated),
+            gather_equations(NewVarNames, Vars0, Bindings),
+            maplist(\Term^Vs^term_variables_under_max_depth(Term, 22, Vs), Bindings, BindingVars),
+            append([ResGoalVars | BindingVars], Vars3),
+            term_variables(Vars3, Vars4), % deduplicate vars of Vars1 but preserve their order.
+            charsio:extend_var_list(Vars4, VarNames, NewVarNames1, fabricated),
+            (   Pending == true ->
+                call(Callback_1, pending(leaf_answer(Bindings, ResGoals, NewVarNames1)))
+            ;   call(Callback_1, final(leaf_answer(Bindings, ResGoals, NewVarNames1)))
+            )
+        )
+    ).
+run_query_term(_, _, Callback_1) :-
+    % If the whole query failed or we didn't cut in the previous definition of
+    % run_query_term/3 (which means  we are still in the query but it has failed)
+    % then we get here so we have a (tail) false.
+    call(Callback_1, final(false)).
+
+
+submit_query_and_print_results2(QueryTerm, VarNames) :-
+    bb_put('$answer_count', 0),
+    bb_put('$report_all', false),
+    bb_put('$report_n_more', 0),
+    catch(
+        run_query_term(QueryTerm, VarNames, toplevel_query_callback),
+        '$stop_query',
+        true
+    ).
+
+handle_first_answer :-
+    (   bb_get('$answer_count', 0) ->
+        write('   ')
+    ;   true
+    ).
+
+increment_answer_count :-
+    bb_get('$answer_count', Count0),
+    Count is Count0 + 1,
+    bb_put('$answer_count', Count).
+
+toplevel_query_callback(pending(LeafAnswer)) :-
+    handle_first_answer,
+    increment_answer_count,
+    show_leaf_answer(LeafAnswer, []),
+    read_input2(LeafAnswer).
+toplevel_query_callback(final(LeafAnswer)) :-
+    (   subsumes_term(exception(_), LeafAnswer) ->
+        exception(Exception) = LeafAnswer,
+        print_exception(Exception)
+    ;   handle_first_answer,
+        increment_answer_count,
+        show_leaf_answer(LeafAnswer, []),
+        write('.'), nl
```

Comment:
No need to quote the atom&nbsp;`.` here.

Reply:
I knew you would point this out. I think quoting it is clearer, as it's a character that is also used for special things (ending clauses) so it can get a bit confusing. A bit related: why does `write_canonical/1` print lists with the quoted dot?

`´´prolog
?- write_canonical([1,2,3]), nl.
'.'(1,'.'(2,'.'(3,[])))
   true.
?- X = .(1,.(2,.(3,[]))).
   % This also works.
   X = [1,2,3].
```

This seems to just make serializing lists (and especially strings of chars) even more inefficient. Is it for readability? If it is, that's exactly my point, but I don't think this should be an big concern for `write_canonical/1` in specific.

Reply:
(my guess is it's the ISO standard). :popcorn:

Reply:
[7.10.5 d](http://www.complang.tuwien.ac.at/ulrich/iso-prolog/draft-7.10.5#d) requires the quoting of a single dot atom with an effective write option `quoted(true)` which means that [`writeq/1`](https://www.complang.tuwien.ac.at/ulrich/iso-prolog/conformity_testing#32) as well as `write_canonical/1` have to quote it in any context. Note that a single dot is very brittle otherwise. Add a space at the end and the dot-atom becomes an end-token.

There [are](https://stackoverflow.com/q/59821549/772868) faster and more space efficient ways of writing. But as long as #1714 and #16 are open, it does not make sense to make things faster here. For conformity, this very verbose format has to be used. Yes at least 7n+2 compared to at least n+2 with double quoted lists.  It could be that also reading could be more efficient here, [but this is not that sure](https://stackoverflow.com/q/70238387/772868).